### ISSUE TITLE
Locking around Symtab::mod_lookup

### ISSUE THREAD

**Issue by hainest** [dyninst/dyninst#1446]
@mwkrentel @jmellorcrummey asked about need of locking in `Symtab::findModuleByOffset`. It looks to be needed because `Symtab::mod_lookup` lazily assigns the `mod_lookup_` member which is an `IBSTree<ModRange>`. Constructing an `IBSTree` appears to be inexpensive: IBSTree.h:258 because an `IBSNode<T>` just has a bunch of primitives and an enum. I see no reason why `mod_lookup_` couldn't be initialized in the constructor and its existence be a class invariant.

It also looks like the same could be done with `func_lookup`. We should survey the whole class.

@kupsch @bigtrak

**Comment by jmellorcrummey** [comment]
@hainest keep in mind that based on our exploration this morning, the implementation of `mod_lookup_`  as `IBSNode<T>` probably needs to be replaced by `IBSNode_fast<T>`. I think that only `IBSNode_fast<T>` has the reader/writer locking.

**Comment by kupsch** [comment]
@hainest, I agree, both `mod_lookup_` and `func_lookup` could be made non-pointer member that are constructed when Symtab is constructed with changing their types to `IBSNode_fast<>`.  `mdForDebugInfo` appears to unused and can be removed. `obj_private` seems like it can potentially be dereferenced with a null value if a `Symtab` is copy constructed perhaps the copy constructor should be deleted.

**Comment by mwkrentel** [comment]
I wrote up the effect of this issue on hpcstruct in:

https://gitlab.com/hpctoolkit/hpctoolkit/-/issues/776

I can confirm that the issue of line map info exists but not in the
primary symtab function module does exist.  It's not real frequent,
but does exist.  For example, in the infamous (static) nwchem binary:

```
0x5d3f200  in  '__parse_one_specmb'

0000000005d3f190 <__parse_one_specmb>:
__parse_one_specmb():
/usr/src/packages/BUILD/glibc-2.11.3/stdio-common/printf-parsemb.c:65
...
/usr/src/packages/BUILD/glibc-2.11.3/stdio-common/printf-parse.h:79
 5d3f200:       8d 04 92                lea    (%rdx,%rdx,4),%eax
 ```

I don't have a clear sense of what conditions trigger this event.
But it might have to do with embedding a `.a` file (with line map info)
into another binary without it, or some such, I don't know.

Anyway, aside from the locking, it would be useful if there was a
version of `findModuleByOffset()` for line map info that only looked
in modules that actually had some line map info.  In the bad case of
`libmkl_avx512.so`, nothing has line map info.  So, the initial hint
fails, then it makes a fruitless search of all other modules which
also fails because none of them have any line map info.
